gazebo_ros_control/src/gazebo_ros_control_plugin.cpp

```diff
@@ -0,0 +1,509 @@
+    //controller->get_lifecycle_node()->set_parameter(rclcpp::Parameter("use_sim_time", true));
+    //controller2->get_lifecycle_node()->set_parameter(rclcpp::Parameter("use_sim_time", true));
+    //controller2->get_lifecycle_node()->set_parameter(rclcpp::Parameter("use_sim_time", true));
+    controller_->get_lifecycle_node()->declare_parameter("period", 0.0);
+    if (controller2_)
+      controller2_->get_lifecycle_node()->declare_parameter("period", 0.0);
+
+    auto spin = [this]()
+    {
+      while (rclcpp::ok())
+        executor_->spin_once();
+    };
+    thread_executor_spin_ = std::thread(spin);
+
+    if (controller_manager_->configure() != controller_interface::CONTROLLER_INTERFACE_RET_SUCCESS) {
+      RCLCPP_ERROR(rclcpp::get_logger("cm"), "failed to configure");
+    }
+    if (controller_manager_->activate() != controller_interface::CONTROLLER_INTERFACE_RET_SUCCESS) {
+      RCLCPP_ERROR(rclcpp::get_logger("cm"), "failed to activate");
+    }
+
+#endif
+
+    // Listen to the update event. This event is broadcast every simulation iteration.
+    update_connection_ =
+      gazebo::event::Events::ConnectWorldUpdateBegin
+      (boost::bind(&GazeboRosControlPlugin::Update, this));
+
+  }
+  catch(pluginlib::LibraryLoadException &ex)
+  {
+    RCLCPP_FATAL_STREAM(rclcpp::get_logger("gazebo_ros_control"),"Failed to create robot simulation interface loader: "<<ex.what());
+  }
+
+  RCLCPP_INFO(rclcpp::get_logger("gazebo_ros_control"), "Loaded gazebo_ros_control.");
+}
+
+// Called by the world update start event
+void GazeboRosControlPlugin::Update()
+{
+  // Get the simulation time and period
+#if GAZEBO_MAJOR_VERSION >= 8
+  gazebo::common::Time gz_time_now = parent_model_->GetWorld()->SimTime();
+#else
+  gazebo::common::Time gz_time_now = parent_model_->GetWorld()->GetSimTime();
+#endif
+  rclcpp::Time sim_time_ros(gz_time_now.sec, gz_time_now.nsec);
+  rclcpp::Duration sim_period = sim_time_ros - last_update_sim_time_ros_;
+
+  robot_hw_sim_->eStopActive(e_stop_active_);
+  //executor_->spin_once();
+
+  // Check if we should update the controllers
+  if(sim_period >= control_period_) {
+    // Store this simulation time
+    last_update_sim_time_ros_ = sim_time_ros;
+
+    // Update the robot simulation with the state of the gazebo model
+    robot_hw_sim_->readSim(sim_time_ros, sim_period);
+
+    // Compute the controller commands
+    bool reset_ctrlrs;
+    if (e_stop_active_)
+    {
+      reset_ctrlrs = false;
+      last_e_stop_active_ = true;
+    }
+    else
+    {
+      if (last_e_stop_active_)
+      {
+        reset_ctrlrs = true;
+        last_e_stop_active_ = false;
+      }
+      else
+      {
+        reset_ctrlrs = false;
+      }
+    }
+
+    controller_->get_lifecycle_node()->set_parameter(rclcpp::Parameter("period", sim_period.seconds()));
+    if (controller2_)
+      controller2_->get_lifecycle_node()->set_parameter(rclcpp::Parameter("period", sim_period.seconds()));
```

Comment:
In ROS1, iirc each controller derives off Gazebo::Node, which supplies the "period" parameter automatically. In ROS2, each controller derives off LifecycleNode and the functionality interface with Gazebo::Node hasn't been done yet. Porters should take this into account.